File: src/codex_team/protocol.py
"""Wire protocol: newline-delimited JSON."""

from __future__ import annotations

import json
from typing import Any, Literal

from pydantic import BaseModel, Field, ValidationError

from codex_team.errors import InvalidRequest


class Request(BaseModel):
    id: str
    cmd: str
    params: dict[str, Any] = Field(default_factory=dict)


class Response(BaseModel):
    id: str
    ok: bool
    data: dict[str, Any] | None = None
    error: dict[str, Any] | None = None


class StreamEvent(BaseModel):
    kind: Literal["event"] = "event"
    stream: Literal["events", "watchdog"]
    seq: int
    payload: dict[str, Any]


def encode(msg: Request | Response | StreamEvent) -> str:
    return msg.model_dump_json(exclude_none=True) + "\n"


def _validate(model: type[BaseModel], data: dict[str, Any]) -> Any:
    try:
        return model.model_validate(data)
    except ValidationError as exc:
        raise InvalidRequest(f"Schema error: {exc}") from exc


def decode_message(line: str) -> Request | Response | StreamEvent:
    payload = line.rstrip("\n")
    if not payload:
        raise InvalidRequest("Empty line")
    try:
        data = json.loads(payload)
    except json.JSONDecodeError as exc:
        raise InvalidRequest(f"Bad JSON: {exc}") from exc
    if not isinstance(data, dict):
        raise InvalidRequest("Message is not an object")
    if "cmd" in data and "id" in data:
        return _validate(Request, data)
    if "ok" in data and "id" in data:
        return _validate(Response, data)
    if data.get("kind") == "event":
        return _validate(StreamEvent, data)
    raise InvalidRequest("Unknown message shape")

File: plugins/codex-team/src/protocol.ts
import { InvalidRequest, ProtocolError, WireError } from "./errors";

export interface RequestMessage {
  v: number;
  id: string;
  cmd: string;
  workspace: string;
  clientId: string | null;
  allWorkspaces: boolean;
  params: Record<string, unknown>;
}

export interface ResponseMessage {
  v?: number;
  id: string;
  ok: boolean;
  workspace?: string;
  data?: Record<string, unknown>;
  error?: WireError;
}

export interface StreamEventMessage {
  kind: "event";
  stream: "events" | "watchdog";
  seq: number;
  payload: Record<string, unknown>;
}

export function encodeMessage(message: RequestMessage | ResponseMessage | StreamEventMessage): string {
  return `${JSON.stringify(message)}\n`;
}

export function decodeRequest(line: string): RequestMessage {
  const payload = decodeJsonObject(line);
  if (typeof payload.id !== "string" || typeof payload.cmd !== "string") {
    throw new InvalidRequest("request must include string id and cmd");
  }
  const v = Number(payload.v);
  if (v !== 2) {
    throw new ProtocolError("request must include protocol v=2");
  }
  if (typeof payload.workspace !== "string" || !payload.workspace) {
    throw new ProtocolError("request must include workspace");
  }
  return {
    v,
    id: payload.id,
    cmd: payload.cmd,
    workspace: payload.workspace,
    clientId: payload.clientId == null ? null : String(payload.clientId),
    allWorkspaces: Boolean(payload.allWorkspaces),
    params: isObject(payload.params) ? payload.params : {},
  };
}

export function decodeJsonObject(line: string): Record<string, unknown> {
  const payload = line.replace(/\n$/, "");
  if (!payload) {
    throw new InvalidRequest("empty line");
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(payload);
  } catch (error) {
    throw new InvalidRequest(`bad JSON: ${(error as Error).message}`);
  }
  if (!isObject(parsed)) {
    throw new InvalidRequest("message is not an object");
  }
  return parsed;
}

export function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
